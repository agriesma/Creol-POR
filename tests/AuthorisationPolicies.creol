class SAuth implements Auth
  begin
    var gr: Agent
    op run == gr := null
    op delay == await (gr = null)
  with Any
    op grant(in x:Agent) == delay(); gr := x
    op revoke(in x:Agent) == if gr = x then gr := null fi
    op auth(in x:Agent) == await (gr = x)
  end 

class MAuth(gr: Agent) implements Auth
begin 
  var gr: Agent
  op run == gr := null
  op delay == await (gr = null)
with Any
  op grant(in x:Agent) == delay(); gr := x
  op revoke(in x:Agent) == if gr = x then gr := null fi
  op auth(in x:Agent) == await (gr = x)
end 

class HAuth implements High 
            inherits SAuth, DB
begin with Agent
  op openH(out ok:Bool) == 
  	await clear(caller;ok);
  	if ok then grant<<Sauth(caller) fi
  op access(in key:int; out y:Data) == 
  	auth<<SAuth(caller);
  	await access@DB(key, true; y)
  op closeH == revoke(caller)
end

class LAuth implements Low 
            inherits MAuth, DB
begin with Agent
  op openL == -- <denne er gal!!!!!!!!! skal ikke kalle clear(..)!!> 
  	await clear(caller;ok);
  	if ok then grant<<MAuth(caller) fi
  op access(in key:int; out y:Data) == 
  	auth<<MAuth(caller);
  	await access@DB(key, false; y)
  op closeL == revoke(caller)
end

interface HighLow inherits High, Low
begin
end

class HLAuth implements HighLow
             inherits LAuth, HAuth
begin with Agent
  op access(in key:int; out y:Data) == if caller=gr@SAuth then access2@HAuth(key, caller;y) else access2@LAuth(key, caller;y) fi
end

class TestAgent(db:HLAuth) implements Agent
begin
	op run ==
		var result:String;
		var ok:bool;
		db.openH(;ok);
		if ok
		then 
			db.access(123; result);
			-- print "Result 1: " + result;
			db.closeH()
		fi;
		db.openL();
		db.access(123; result);
		-- print "Result 2: " + result;
		db.access(456; result);
		-- print "Result 3: " + result;
		db.closeL()
end

class DB 
begin 
	op access(in key:int, b:boolean; out y:Data) ==
		if (key = 123) 
		then if b
			 then y := "123 is customer Ole Hansen. Phone no: 90807060." 
			 else y := "123 is customer Ole Hansen."
			 fi
		else y := "Unknown customer."
		fi
end
