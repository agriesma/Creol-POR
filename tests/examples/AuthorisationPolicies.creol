interface Agent
begin
end

interface Auth
begin
with Any
  op grant(in x: Agent)
  op revoke(in x: Agent)
  op auth(in x: Agent)
end

class SAuth
  implements Auth
begin
  var gr: Agent
  op init == skip
  op run == gr := null
  op delay == await (gr = null)
with Any
  op grant(in x:Agent) == delay(;); gr := x
  op revoke(in x:Agent) == if gr = x then gr := null end
  op auth(in x:Agent) == await (gr = x)
end 

class MAuth(gr: Agent)
  implements Auth
begin 
  var gr: Agent
  op init == skip
  op run == gr := null
  op delay == await (gr = null)
with Any
  op grant(in x:Agent) == delay(;); gr := x
  op revoke(in x:Agent) == if gr = x then gr := null end
  op auth(in x:Agent) == await (gr = x)
end

interface High
begin
with Agent
  op openH(out ok: Bool)
  op access(in k: Int; out y: Data)
  op closeH
end

interface Low
begin
with Agent
  op openL(out ok: Bool)
  op access(in k: Int; out y: Data)
  op closeL
end

class Test
  implements High 
  inherits SAuth, DB
begin
  op clear(in x: Agent; out r: Bool) == r := true

with Agent
  op openH(out ok:Bool) == 
    await clear(caller; ok);
    if ok then grant:>SAuth(caller; ok) end
  op access(in key:Int; out y:Data) == 
    auth:>SAuth(caller;); await access<:DB(key, true; y)
  op closeH == revoke(caller;)
end

class LAuth implements Low inherits MAuth, DB
begin with Agent
  op openL == // <denne er gal!!!!!!!!! skal ikke kalle clear(..)!!> 
  	await clear(caller; ok);
  	if ok then grant<:MAuth(caller;) end
  op access(in key:Int; out y:Data) == 
  	auth<:MAuth(caller;);
  	await access:>DB(key, false; y)
  op closeL == revoke(caller;)
end

interface HighLow inherits High, Low
begin
end

class HLAuth implements HighLow inherits LAuth, HAuth
begin with Agent
  op access(in key:Int; out y:Data) ==
    if caller = gr@SAuth then
      access2:>HAuth(key, caller;y)
    else
      access2:>LAuth(key, caller;y)
    end
end

class TestAgent(db:HLAuth) implements Agent
begin
  op run == var result: String; var ok: Bool;
    db.openH(;ok); if ok then db.access(123; result); db.closeH(;) end;
    db.openL(;); db.access(123; result); db.access(456; result);
    db.closeL(;)
end

class DB 
begin
  op access(in key:Int, b:Bool; out y:Data) ==
    if (key = 123) then
      if b then
        y := "123 is customer Ole Hansen. Phone no: 90807060." 
      else
        y := "123 is customer Ole Hansen."
      end
    else
      y := "Unknown customer."
    end
end
