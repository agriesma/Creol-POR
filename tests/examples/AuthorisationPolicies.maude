load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "SAuth" : Cl | Inh: noInh, Par: noVid, Att: "gr" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: "gr" ::= null ; return ( emp ) > *
  < "delay" : Mtdname | Param: noVid, Latt: noSubst, Code: ( await "=" ( "gr" # null ) ) ; return ( emp ) > *
  < "grant" : Mtdname | Param: "x", Latt: "x" |-> null, Code: ( "sync.0" ! "this" . "delay" (emp ) )  ; ( "sync.0" ? ( noVid ) )  ; "gr" ::= "x" ; return ( emp ) > *
  < "revoke" : Mtdname | Param: "x", Latt: "x" |-> null, Code: if "=" ( "gr" # "x" ) th "gr" ::= null el skip fi ; return ( emp ) > *
  < "auth" : Mtdname | Param: "x", Latt: "x" |-> null, Code: ( await "=" ( "gr" # "x" ) ) ; return ( emp ) >, Ocnt: 0 >

< "MAuth" : Cl | Inh: noInh, Par: "gr", Att: "gr" |-> null , "gr" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: "gr" ::= null ; return ( emp ) > *
  < "delay" : Mtdname | Param: noVid, Latt: noSubst, Code: ( await "=" ( "gr" # null ) ) ; return ( emp ) > *
  < "grant" : Mtdname | Param: "x", Latt: "x" |-> null, Code: ( "sync.0" ! "this" . "delay" (emp ) )  ; ( "sync.0" ? ( noVid ) )  ; "gr" ::= "x" ; return ( emp ) > *
  < "revoke" : Mtdname | Param: "x", Latt: "x" |-> null, Code: if "=" ( "gr" # "x" ) th "gr" ::= null el skip fi ; return ( emp ) > *
  < "auth" : Mtdname | Param: "x", Latt: "x" |-> null, Code: ( await "=" ( "gr" # "x" ) ) ; return ( emp ) >, Ocnt: 0 >

< "Test" : Cl | Inh: "SAuth" < emp >  ## "DB" < emp > , Par: noVid, Att: noSubst, Mtds: 
  < "openH" : Mtdname | Param: noVid, Latt: "ok" |-> null, Code: ( "sync.0" ! "this" . "clear" ("caller" ) )  ; ( await ( "sync.0" ?? ) ) ; ( "sync.0" ? ( "ok" ) )  ; if "ok" th ( "sync.1" ! "grant" @ "Sauth" ( "caller" ) )  ; ( "sync.1" ? ( noVid ) )  el skip fi ; return ( "ok" ) > *
  < "access" : Mtdname | Param: "key", Latt: "key" |-> null , "y" |-> null, Code: ( "sync.0" ! "auth" @ "SAuth" ( "caller" ) )  ; ( "sync.0" ? ( noVid ) )  ; ( "sync.1" ! "access" @ "Test" << "DB" ( "key" # bool(true) ) )  ; ( await ( "sync.1" ?? ) ) ; ( "sync.1" ? ( "y" ) )  ; return ( "y" ) > *
  < "closeH" : Mtdname | Param: noVid, Latt: noSubst, Code: ( "sync.0" ! "this" . "revoke" ("caller" ) )  ; ( "sync.0" ? ( noVid ) )  ; return ( emp ) >, Ocnt: 0 >

< "LAuth" : Cl | Inh: "MAuth" < emp >  ## "DB" < emp > , Par: noVid, Att: noSubst, Mtds: 
  < "openL" : Mtdname | Param: noVid, Latt: noSubst, Code: ( "sync.0" ! "this" . "clear" ("caller" ) )  ; ( await ( "sync.0" ?? ) ) ; ( "sync.0" ? ( "ok" ) )  ; if "ok" th ( "sync.1" ! "grant" @ "LAuth" << "MAuth" ( "caller" ) )  ; ( "sync.1" ? ( noVid ) )  el skip fi ; return ( emp ) > *
  < "access" : Mtdname | Param: "key", Latt: "key" |-> null , "y" |-> null, Code: ( "sync.0" ! "auth" @ "LAuth" << "MAuth" ( "caller" ) )  ; ( "sync.0" ? ( noVid ) )  ; ( "sync.1" ! "access" @ "DB" ( "key" # bool(false) ) )  ; ( await ( "sync.1" ?? ) ) ; ( "sync.1" ? ( "y" ) )  ; return ( "y" ) > *
  < "closeL" : Mtdname | Param: noVid, Latt: noSubst, Code: ( "sync.0" ! "this" . "revoke" ("caller" ) )  ; ( "sync.0" ? ( noVid ) )  ; return ( emp ) >, Ocnt: 0 >

< "HLAuth" : Cl | Inh: "LAuth" < emp >  ## "HAuth" < emp > , Par: noVid, Att: noSubst, Mtds: 
  < "access" : Mtdname | Param: "key", Latt: "key" |-> null , "y" |-> null, Code: if "=" ( "caller" # ( "gr" @@ "SAuth" ) ) th ( "sync.1" ! "access2" @ "HAuth" ( "key" # "caller" ) )  ; ( "sync.1" ? ( "y" ) )  el ( "sync.0" ! "access2" @ "LAuth" ( "key" # "caller" ) )  ; ( "sync.0" ? ( "y" ) )  fi ; return ( "y" ) >, Ocnt: 0 >

< "TestAgent" : Cl | Inh: noInh, Par: "db", Att: "db" |-> null, Mtds: 
  < "run" : Mtdname | Param: noVid, Latt: "result" |-> null , "ok" |-> null, Code: "sync.0" ! "db" . "openH" ( emp ) ; ( "sync.0" ? ( "ok" ) )  ; if "ok" th "sync.1" ! "db" . "access" ( int(123) ) ; ( "sync.1" ? ( "result" ) )  ; "sync.2" ! "db" . "closeH" ( emp ) ; ( "sync.2" ? ( noVid ) )  el skip fi ; "sync.3" ! "db" . "openL" ( emp ) ; ( "sync.3" ? ( noVid ) )  ; "sync.4" ! "db" . "access" ( int(123) ) ; ( "sync.4" ? ( "result" ) )  ; "sync.5" ! "db" . "access" ( int(456) ) ; ( "sync.5" ? ( "result" ) )  ; "sync.6" ! "db" . "closeL" ( emp ) ; ( "sync.6" ? ( noVid ) )  ; return ( emp ) >, Ocnt: 0 >

< "DB" : Cl | Inh: noInh, Par: noVid, Att: noSubst, Mtds: 
  < "access" : Mtdname | Param: "key" , "b", Latt: "key" |-> null , "b" |-> null , "y" |-> null, Code: if "=" ( "key" # int(123) ) th if "b" th "y" ::= str("123 is customer Ole Hansen. Phone no: 90807060.") el "y" ::= str("123 is customer Ole Hansen.") fi el "y" ::= str("Unknown customer.") fi ; return ( "y" ) >, Ocnt: 0 >

.
endm
