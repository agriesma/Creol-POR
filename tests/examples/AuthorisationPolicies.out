==========================================
reduce in PROGRAM : init .
result Configuration: init
Bye.
